deepspeed/sequence/layer.py

```diff
@@ -155,7 +155,7 @@ def single_all_to_all(input, scatter_idx, gather_idx, batch_dim_idx, group, asyn
     # we only need num_heads once
     num_heads = input.shape[2]
 
-    if get_num_kv_heads() is not None or num_heads % seq_world_size != 0:
+    if get_num_kv_heads() is not None or (num_heads % seq_world_size != 0 and gather_idx < 2):
```

Comment:
By the way, can we use scatter_idx for the judgment here? This way, we can stay consistent with other parts by uniformly using scatter_idx to determine whether the current state is ```total_head``` or ```total_seq_len```

Reply:
Ditto @inkcherry! Future refactor will have a better naming to differentiate between N/p (pre/post-attention) and h/p (attention block)

Reply:
Hi I agree, scatter_idx booleans are confusing. Maybe I can share my opinion on the refactoring on Wed if thats ok.

Reply:
Hey, I just made a quick change to use `scatter_idx`.